geth-utils/gethutil/mpt/witness/branch.go

```diff
@@ -149,25 +150,30 @@ func prepareBranchNode(branch1, branch2, extNode1, extNode2, extListRlpBytes []b
 	return node
 }
 
-// getDriftedPosition returns the position in branch to which the leaf drifted because another
-// leaf has been added to the same slot. This information is stored into a branch init row.
-func getDriftedPosition(leafKeyRow []byte, numberOfNibbles int) byte {
+// getNibbles returns the nibbles of the leaf or extension node.
+func getNibbles(leafKeyRow []byte) []byte {
 	var nibbles []byte
 	if leafKeyRow[0] != 248 {
-		keyLen := int(leafKeyRow[1] - 128)
-		if (leafKeyRow[2] != 32) && (leafKeyRow[2] != 0) { // second term is for extension node
-			if leafKeyRow[2] < 32 { // extension node
-				nibbles = append(nibbles, leafKeyRow[2]-16)
-			} else { // leaf
-				nibbles = append(nibbles, leafKeyRow[2]-48)
+		var keyLen int
+		if leafKeyRow[1] > 128 {
+			keyLen = int(leafKeyRow[1] - 128)
```

Comment:
A question about the `keyLen`. 
In this `if` condition, `leafKeyRow` is not 248, let's say it's 249, then the length of the payload should be `leafKeyRow[1]*256 + leafKeyRow[2]`. I don't get why the `keyLen` here is `int(leafKeyRow[1] - 128)`. Besides, `leafKeyRow[2]` in below (L160) should be also part of the length.

I also found that the `keyLen` in the `else` case in L178 is using `leafKeyRow[2] ` instead of `leafKeyRow[1]`. I guess the implementation in `if` and `else` should be swithced?

Reply:
The reason this might seem confusing is because for the account and storage leaves I think the length is never big enough for the first byte to be `249`, it's at most `248`.

It would be more clear if the first condition would say `leafKeyRow[0] < 248` instead of `leafKeyRow[0] != 248` (and the second block would need to say `leafKeyRow[0] == 248`).

The first if block handles the cases when `leafKeyRow[0] < 248`, the second block handles the cases for `leafKeyRow[0] == 248`. It works ok for account and storage leaves, but not for the transaction leaves which can be very long and can have the first byte `249`. Another block is needed for these cases.